src/main/java/io/dgraph/DgraphAsyncClient.java

```diff
@@ -221,29 +221,14 @@ protected <T> CompletableFuture<T> runWithRetries(
             throw new RuntimeException(e);
           } catch (ExecutionException e) {
             if (ExceptionUtil.isJwtExpired(e.getCause())) {
-              try {
-                // retry the login
-                retryLogin().get();
-                // retry the supplied logic
-                return ctxCallable.call().get();
-              } catch (InterruptedException ie) {
-                LOG.error("The retried " + operation + " got interrupted:", ie);
-                throw new RuntimeException(ie);
-              } catch (ExecutionException ie) {
-                LOG.error("The retried " + operation + " encounters an execution exception:", ie);
-                throw new RuntimeException(ie);
-              } catch (Exception ie) {
-                LOG.error("The retried " + operation + " encounters a completion exception:", ie);
-                throw new CompletionException(ie);
-              }
+              return retryLoginAndExecute(ctxCallable, operation);
             } else if (e.getCause() instanceof StatusRuntimeException) {
               StatusRuntimeException ex1 = (StatusRuntimeException) e.getCause();
               Status.Code code = ex1.getStatus().getCode();
-              String desc = ex1.getStatus().getDescription();
-
               if (code.equals(Status.Code.ABORTED)
                   || code.equals(Status.Code.FAILED_PRECONDITION)) {
-                throw new CompletionException(new TxnConflictException(desc));
+                return executeWithExponentialBackoff(
```

Comment:
We should provide the user the option to retry the transaction rather than forcing one. In some use-cases like BnF, it may not be desirable to retry all transactions that failed due to a conflict.